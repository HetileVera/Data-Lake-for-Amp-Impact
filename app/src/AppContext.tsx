import React, { createContext, useContext } from "react";
import isEmpty from "lodash/isEmpty";
import throttle from "lodash/throttle";

import { AppStore, IAppStore } from "./models/AppStore";
import { IInstallation } from "./models/Installation";
import { isLocalDev, websiteUrlOrigin, assetWebsiteUrl, secretKey } from "./helpers/settings";
import { onSnapshot } from "mobx-state-tree";

// Create an instance of the root store (a.k.a appStore)
const appStore = AppStore.create({});

// Store state to local storage
const storeLocally = throttle(() => appStore.saveLocal(), 5000);

// Store state in S3 when applicable
const storeRemotely = throttle(() => appStore.saveRemote(), 30000);

// The initialization function for the root store
export function initializeStore() {
  if (!isEmpty(secretKey)) console.warn("Access key is bundled!");

  if (isLocalDev) {
    console.log(`isLocalDev '${isLocalDev}'`);
    console.log(`websiteUrlOrigin '${websiteUrlOrigin}'`);
    console.log(`assetWebsiteUrl '${assetWebsiteUrl}'`);

    if (window.location.host !== "localhost:3000") {
      console.warn("You should run the app on localhost!");

      throw new Error("App should run on localhost for development mode!");
    }
  }

  appStore.loadLocal();
  onSnapshot(appStore, (snapshot) => {
    storeLocally();
    storeRemotely();
  });
}

// Create the store context
export const StoreContext = createContext<IAppStore | undefined>(undefined);

// Create the store provider component
export const StoreProvider: React.FunctionComponent = ({ children }) => {
  return <StoreContext.Provider value={appStore}>{children}</StoreContext.Provider>;
};

// useStore hook
export function useStore(): IAppStore {
  const store = useContext(StoreContext);

  if (!store) {
    throw new Error("useStore() must be used within a StoreProvider");
  }

  return store;
}

// useInstallation hook
export function useInstallation(): IInstallation {
  const store = useStore();
  const installation = store.installation;

  if (!installation) {
    throw new Error("An installation instance was not created yet");
  }

  return installation;
}
