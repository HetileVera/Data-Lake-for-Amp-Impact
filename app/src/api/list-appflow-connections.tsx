import { AppflowClient, ConnectorType, DescribeConnectorProfilesCommand, DescribeConnectorProfilesCommandInput } from "@aws-sdk/client-appflow";

/**
 * Returns a list of all the connection names for the Salesforce connector.
 */
export async function listAppflowConnections({
  accessKey,
  secretKey,
  region,
}: {
  accessKey: string;
  secretKey: string;
  region: string;
}): Promise<string[]> {
  const appFlowClient = new AppflowClient({
    region,
    credentials: {
      accessKeyId: accessKey,
      secretAccessKey: secretKey,
    },
  });

  let params: DescribeConnectorProfilesCommandInput = { connectorType: ConnectorType.SALESFORCE, maxResults: 100 };
  const names: string[] = [];

  do {
    const data = await appFlowClient.send(new DescribeConnectorProfilesCommand(params));
    params.nextToken = data.nextToken;
    names.push(...((data.connectorProfileDetails ?? []).map((connector) => connector.connectorProfileName as string)));
  } while (params.nextToken);

  return names;
}
