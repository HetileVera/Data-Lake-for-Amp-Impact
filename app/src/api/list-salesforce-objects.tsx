import { AppflowClient, ConnectorType, ListConnectorEntitiesCommand, ListConnectorEntitiesCommandInput } from "@aws-sdk/client-appflow";
import _get from "lodash/get";

export interface IRawSalesforceObject {
  name: string;
  label: string;
}

/**
 * Returns a list of all objects in Salesforce
 */
export async function listSalesforceObjects({
  accessKey,
  secretKey,
  region,
  connectionName,
}: {
  accessKey: string;
  secretKey: string;
  region: string;
  connectionName: string;
}): Promise<IRawSalesforceObject[]> {
  const appFlowClient = new AppflowClient({
    region,
    credentials: {
      accessKeyId: accessKey,
      secretAccessKey: secretKey,
    },
  });

  const params: ListConnectorEntitiesCommandInput = { connectorType: ConnectorType.SALESFORCE, connectorProfileName: connectionName };
  const data = await appFlowClient.send(new ListConnectorEntitiesCommand(params));
  const objects = _get(data, "connectorEntityMap.Objects", []);

  return objects.map((obj: { name: string; label: string }) => ({ name: obj.name, label: obj.label }));
}
