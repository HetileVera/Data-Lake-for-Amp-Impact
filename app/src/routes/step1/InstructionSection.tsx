import React from "react";
import { Box, Divider, Stack, Link as ChakraLink, UnorderedList, ListItem } from "@chakra-ui/react";
import { ExternalLinkIcon } from "@chakra-ui/icons";

import awsConnectImage01 from "images/aws-connect-01.png";
import awsConnectImage02 from "images/aws-connect-02.png";

import { ClickableImage } from "components/ClickableImage";

export const InstructionSection = () => {
  return (
    <Box bg="orange.75" p={6} borderRadius="lg" border="1px" borderColor="orange.100">
      <Stack direction={{ base: "column", md: "row" }} spacing="30px" pt={6} pb={6}>
        <Box>
          <ClickableImage w={{ base: "full", md: "300px" }} src={awsConnectImage01} title="AWS Console Login Page" />
        </Box>
        <Box>
          <UnorderedList ml={5} mt={0} mb={3} color="orange.900">
            <ListItem>
              Log in to the &nbsp;
              <ChakraLink href="https://aws.amazon.com/console/" isExternal color="orange.600">
                AWS console <ExternalLinkIcon mx="2px" />
              </ChakraLink>
              &nbsp; using an admin user, not the root user.
            </ListItem>
          </UnorderedList>
          <Box fontSize="sm" mt={4} bg="orange.100" p={3} borderRadius="lg">
            Remember to log in using an admin user and not the root user
          </Box>
        </Box>
      </Stack>

      <Divider borderColor="orange.200" />

      <Stack direction={{ base: "column", md: "row" }} spacing="30px" mt={6} mb={0}>
        <Box>
          <ClickableImage w={{ base: "full", md: "300px" }} src={awsConnectImage02} title="AWS Console IAM Page" />
        </Box>
        <Box>
          <UnorderedList ml={5} mt={0} mb={0} color="orange.900">
            <ListItem>
              Then, head over to the{" "}
              <ChakraLink href="https://console.aws.amazon.com/iamv2/home#/users" isExternal color="orange.600">
                Identity and Management section (IAM) <ExternalLinkIcon mx="2px" />
              </ChakraLink>
            </ListItem>
            <ListItem mt={4}>
              Pick the user name that you used to log in. The user should have admin permissions. Then, select the Security Credentials tab.
            </ListItem>
            <ListItem mt={4}>
              Click on the Create access key button to generate the access keys, you will get a chance to view and download the access keys. You can
              now use the access keys to fill in the information below.
            </ListItem>
          </UnorderedList>
        </Box>
      </Stack>
    </Box>
  );
};
