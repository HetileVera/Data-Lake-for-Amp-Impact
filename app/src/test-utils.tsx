import React from "react";
import { render, RenderOptions } from "@testing-library/react";
import { ChakraProvider } from "@chakra-ui/react";

import { theme } from "themes/orange";

export const CustomChakraProvider = ({ children }: { children: React.ReactNode }) => {
  return (
    <ChakraProvider theme={theme}>
      <div data-testid="test-root">{children}</div>
    </ChakraProvider>
  );
};

const customRender = (ui: React.ReactElement, options?: Omit<RenderOptions, "wrapper">) => render(ui, { wrapper: CustomChakraProvider, ...options });

export * from "@testing-library/react";

export { customRender as render };
