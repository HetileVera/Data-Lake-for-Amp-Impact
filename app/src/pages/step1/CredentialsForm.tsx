import { FC, useState } from "react";
import isEmpty from "lodash/isEmpty";
import { Box, Heading, Button, FormErrorMessage, FormControl, Input, Alert, AlertIcon, InputGroup, InputRightElement } from "@chakra-ui/react";
import { UseFormRegister, FieldErrors, useWatch, Control } from "react-hook-form";

import { useStore } from "AppContext";
import { useColorScheme } from "models/useColorScheme";
import { accessKey, secretKey } from "helpers/settings";

export interface ICredentialsFormInput {
  accessKeyId: string;
  secretAccessKey: string;
  region: string;
}
export interface ICredentialsForm {
  register: UseFormRegister<ICredentialsFormInput>;
  errors: FieldErrors<ICredentialsFormInput>;
  isSubmitting: boolean;
  control: Control<ICredentialsFormInput>;
}

export const CredentialsForm = ({ register, errors, isSubmitting, control }: ICredentialsForm) => {
  const store = useStore();
  const { tone, colorScheme } = useColorScheme();
  const credentials = store.credentials;
  const accessKeyDefaultValue = credentials.accessKey || accessKey;
  const secretKeyDefaultValue = credentials.secretKey || secretKey;
  const secretKeyValue = useWatch({ control, name: "secretAccessKey", defaultValue: secretKeyDefaultValue });
  const [show, setShow] = useState(false);
  const handleClick = () => setShow(!show);

  return (
    <Box p={6} borderRadius="lg" boxShadow="base" bg={tone(50)}>
      <Heading size="sm" pt="5px" pb="20px" color={tone(600)}>
        Access Key ID &amp; Secret Access Key
      </Heading>

      {/* @ts-ignore */}
      <FormControl isInvalid={errors.accessKeyId || errors.secretAccessKey}>
        <Input
          id="accessKeyId"
          placeholder="Access key id"
          defaultValue={accessKeyDefaultValue}
          {...register("accessKeyId", { required: "Required" })}
          focusBorderColor={tone(700)}
          bg={tone(75)}
          disabled={isSubmitting}
        />
        <FormErrorMessage>{errors.accessKeyId && errors.accessKeyId.message}</FormErrorMessage>

        <InputGroup size="md" mt={5}>
          <Input
            pr="4.5rem"
            type={show ? "text" : "password"}
            placeholder="Secret access key"
            id="secretAccessKey"
            defaultValue={secretKeyDefaultValue}
            {...register("secretAccessKey", { required: "Required" })}
            focusBorderColor={tone(700)}
            bg={tone(75)}
            disabled={isSubmitting}
          />
          <InputRightElement width="4.5rem">
            <Button
              h="1.75rem"
              size="sm"
              onClick={handleClick}
              colorScheme={colorScheme}
              variant="outline"
              _focus={{ boxShadow: "none" }}
              isDisabled={isEmpty(secretKeyValue)}
            >
              {show ? "Hide" : "Show"}
            </Button>
          </InputRightElement>
        </InputGroup>

        <FormErrorMessage>{errors.secretAccessKey && errors.secretAccessKey.message}</FormErrorMessage>
        <Alert status="warning" variant="left-accent" mt={5} fontSize="sm">
          <AlertIcon />
          Secret access keys are sensitive information, keep them secured and protected, never post them on public platforms or leave them unsecured
          as this can compromise your account security
        </Alert>
      </FormControl>
    </Box>
  );
};
