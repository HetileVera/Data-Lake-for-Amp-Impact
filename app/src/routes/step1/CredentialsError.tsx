import { FC } from "react";
import { observer } from "mobx-react";
import { Alert, AlertDescription, AlertIcon, AlertTitle, Box, ListItem, UnorderedList } from "@chakra-ui/react";

import { CredentialsValidationErrorCode, CredentialsValidationException } from "../../api/validate-credentials";

export const CredentialsError: FC<{ exception?: CredentialsValidationException | Error }> = observer(({ exception }) => {
  if (!exception) return null;

  let title: string;
  let content: React.ReactElement = <></>;

  if (exception instanceof CredentialsValidationException) {
    title = "A problem with the access keys";
    switch (exception?.code) {
      case CredentialsValidationErrorCode.InvalidKeys:
        content = (
          <Box>
            The keys are invalid. This could be due to one or more of the following reasons:
            <UnorderedList ml={5} mt={2}>
              <ListItem>The keys are incorrect, check that you are using the correct values</ListItem>
              <ListItem>The access keys are deactivated</ListItem>
              <ListItem>The access keys are deleted and therefore no longer valid</ListItem>
            </UnorderedList>
          </Box>
        );
        break;
      case CredentialsValidationErrorCode.NotAdmin:
        content = (
          <Box>
            The keys are valid but they are for a user who does not have admin permissions. Add the <b>AdministratorAccess</b> policy to the user and
            try again.
          </Box>
        );
        break;
      case CredentialsValidationErrorCode.AccountMismatch:
        content = <Box>The keys are for a different AWS account.</Box>;
        break;

      default:
        content = (
          <Box>
            We are unable to validate the keys. This could be due to one or more of the following reasons:
            <UnorderedList ml={5} mt={2} mb={3}>
              <ListItem>A problem with the internet connection</ListItem>
              <ListItem>An intermittent problem while trying to validate the keys</ListItem>
            </UnorderedList>
            Try again in a few minutes
          </Box>
        );
        break;
    }
  } else {
    title = "A problem has been encountered";
    content = (
      <Box>
        We are unable to proceed to the next step. This could be due to one or more of the following reasons:
        <UnorderedList ml={5} mt={2} mb={3}>
          <ListItem>A problem with the internet connection</ListItem>
          <ListItem>An intermittent problem while trying to proceed to the next step</ListItem>
        </UnorderedList>
        Try again in a few minutes
      </Box>
    );
  }

  return (
    <Alert status="error" variant="left-accent" mt={5} color="red.700" alignItems="flex-start" borderColor="red.600">
      <AlertIcon mt="3px" color="red.600" />
      <Box>
        <AlertTitle mr={2}>{title}</AlertTitle>
        <AlertDescription>{content}</AlertDescription>
      </Box>
    </Alert>
  );
});
