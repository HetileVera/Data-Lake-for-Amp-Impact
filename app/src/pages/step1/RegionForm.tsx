import React from "react";
import { FieldErrors, UseFormRegister } from "react-hook-form";
import { Box, FormControl, FormErrorMessage, Heading, Select, Text } from "@chakra-ui/react";

import { ICredentialsFormInput } from "./CredentialsForm";
import { awsRegions } from "../../data/aws-regions";
import { useColorScheme } from "../../models/useColorScheme";

interface IRegionForm {
  register: UseFormRegister<ICredentialsFormInput>;
  errors: FieldErrors;
  isSubmitting: boolean;
  description: string;
  defaultRegion: string;
}

export const RegionForm = ({ register, errors, isSubmitting, description, defaultRegion }: IRegionForm) => {
  const { tone } = useColorScheme();

  return (
    <Box p={6} bg={tone(50)} borderRadius="lg" boxShadow="base">
      <Heading size="sm" pt="5px" pb="5px" color={tone(600)}>
        AWS Region
      </Heading>
      <Text fontSize="sm" color={tone(700)} mb={3}>
        {description}
      </Text>
      {/* @ts-ignore */}
      <FormControl isInvalid={errors.region || errors.region}>
        <Select bg={tone(75)} id="region" {...register("region", { required: "Required" })} defaultValue={defaultRegion} disabled={isSubmitting}>
          {awsRegions.map((item) => (
            <option key={item.name} className={item.name} value={item.name}>
              {item.label}
            </option>
          ))}
        </Select>
        <FormErrorMessage>{errors.region && errors.region.message}</FormErrorMessage>
      </FormControl>
    </Box>
  );
};
