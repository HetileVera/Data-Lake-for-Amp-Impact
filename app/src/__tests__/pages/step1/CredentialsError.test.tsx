import { cleanup, screen } from "@testing-library/react";
import { create } from "react-test-renderer";

import { render, CustomChakraProvider } from "test-utils";

import { StoreProvider } from "AppContext";

import { CredentialsError } from "pages/step1/CredentialsError";
import { CredentialsValidationErrorCode, CredentialsValidationException } from "api/validate-credentials";

describe("Step1 -> CredentialsError component", () => {
  afterEach(cleanup);

  test("snapshot", () => {
    const tree = create(
      <CustomChakraProvider>
        <StoreProvider>
          <CredentialsError />
        </StoreProvider>
      </CustomChakraProvider>
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  test("render properly without an exception", () => {
    render(
      <StoreProvider>
        <CredentialsError />
      </StoreProvider>
    );

    expect(screen.findByRole("body")).toBeTruthy();
    expect(screen.queryByText(/error/i)).not.toBeInTheDocument();
  });

  test("render properly with an invalid key exception", () => {
    const exception = new CredentialsValidationException(CredentialsValidationErrorCode.InvalidKeys);

    render(
      <StoreProvider>
        <CredentialsError exception={exception} />
      </StoreProvider>
    );

    expect(screen.queryByText(/a problem with the access keys/i)).toBeInTheDocument();
    expect(screen.queryByText(/the keys are invalid/i)).toBeInTheDocument();
  });

  test("render properly with a non-admin key exception", () => {
    const exception = new CredentialsValidationException(CredentialsValidationErrorCode.NotAdmin);

    render(
      <StoreProvider>
        <CredentialsError exception={exception} />
      </StoreProvider>
    );

    expect(screen.queryByText(/a problem with the access keys/i)).toBeInTheDocument();
    expect(screen.queryByText(/the keys are valid but they are for a user who does not have admin permissions/i)).toBeInTheDocument();
  });

  test("render properly with a non-admin key exception", () => {
    const exception = new CredentialsValidationException(CredentialsValidationErrorCode.AccountMismatch);

    render(
      <StoreProvider>
        <CredentialsError exception={exception} />
      </StoreProvider>
    );

    expect(screen.queryByText(/a problem with the access keys/i)).toBeInTheDocument();
    expect(screen.queryByText(/the keys are for a different aws account/i)).toBeInTheDocument();
  });

  test("render properly with a default exception", () => {
    const exception = new CredentialsValidationException(CredentialsValidationErrorCode.Unknown);

    render(
      <StoreProvider>
        <CredentialsError exception={exception} />
      </StoreProvider>
    );

    expect(screen.queryByText(/a problem with the access keys/i)).toBeInTheDocument();
    expect(screen.queryByText(/we are unable to validate the keys/i)).toBeInTheDocument();
  });

  test("render properly with an unknown exception", () => {
    const exception = new Error();

    render(
      <StoreProvider>
        <CredentialsError exception={exception} />
      </StoreProvider>
    );

    expect(screen.queryByText(/a problem has been encountered/i)).toBeInTheDocument();
    expect(screen.queryByText(/we are unable to proceed to the next step/i)).toBeInTheDocument();
  });
});
