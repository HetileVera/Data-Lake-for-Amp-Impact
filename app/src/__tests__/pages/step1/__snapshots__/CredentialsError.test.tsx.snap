// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Step1 -> CredentialsError component snapshot 1`] = `<span />`;
